components/sup/src/ctl_gateway/client.rs

```diff
@@ -0,0 +1,251 @@
+//! conn.call(msg).for_each(|reply| {
+//!     match reply.message_id() {
+//!         "ServiceCfg" => {
+//!             let m = reply.parse::<protocols::types::ServiceCfg>().unwrap();
+//!             println!("{}", m.get_default());
+//!         }
+//!         "NetErr" => {
+//!             let m = reply.parse::<protocols::net::NetErr>().unwrap();
+//!             println!("{}", m);
+//!         }
+//!         _ => (),
+//!     }
+//!     Ok(())
+//! })
+//! ```
+
+use std::error;
+use std::fmt;
+use std::io;
+use std::net::SocketAddr;
+
+use futures::sink;
+use futures::prelude::*;
+use protobuf;
+use tokio::net::TcpStream;
+use tokio_io::AsyncRead;
+
+use super::codec::*;
+use protocols;
+use protocols::net::NetErr;
+
+pub type SrvSend = sink::Send<SrvStream>;
+
+/// Error types returned by a [`SrvClient`].
+#[derive(Debug)]
+pub enum SrvClientError {
+    /// The remote server unexpectedly closed the connection.
+    ConnectionClosed,
+    /// Decoding a message from the remote failed.
+    Decode(protobuf::ProtobufError),
+    /// An Os level IO error occurred.
+    Io(io::Error),
+    /// An RPC call to the remote was received but failed.
+    NetErr(NetErr),
+}
+
+impl error::Error for SrvClientError {
+    fn description(&self) -> &str {
+        match *self {
+            SrvClientError::ConnectionClosed => "Connection closed",
+            SrvClientError::Decode(ref err) => err.description(),
+            SrvClientError::Io(ref err) => err.description(),
+            SrvClientError::NetErr(ref err) => err.description(),
+        }
+    }
+}
+
+impl fmt::Display for SrvClientError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let content = match *self {
+            SrvClientError::ConnectionClosed => format!("Connection closed"),
+            SrvClientError::Decode(ref err) => format!("{}", err),
+            SrvClientError::Io(ref err) => format!("{}", err),
+            SrvClientError::NetErr(ref err) => format!("{}", err),
+        };
+        write!(f, "{}", content)
+    }
+}
+
+impl From<NetErr> for SrvClientError {
+    fn from(err: NetErr) -> Self {
+        SrvClientError::NetErr(err)
+    }
+}
+
+impl From<io::Error> for SrvClientError {
+    fn from(err: io::Error) -> Self {
+        SrvClientError::Io(err)
+    }
+}
+
+impl From<protobuf::ProtobufError> for SrvClientError {
+    fn from(err: protobuf::ProtobufError) -> Self {
+        SrvClientError::Decode(err)
+    }
+}
+
+/// Client for connecting and communicating with a server listener which speaks SrvProtocol.
+///
+/// See module doc for usage.
+pub struct SrvClient {
+    /// Sending and receiving framed socket pair for connecting and communicating to a remote
+    /// server.
+    socket: SrvStream,
+    /// Transaction ID counter.
+    ///
+    /// Useful for determining the next transaction ID in sequence embed in
+    /// the next transactional message sent.
+    current_txn: u32,
```

Comment:
Why not make this a `SrvTxn`? It seems like
```rust
        self.current_txn = self.current_txn.next();
```
would provide more safety than
```rust
        self.current_txn = SrvTxn::next_id(self.current_txn);
```
especially if `SrvTxn::new(u32)` is eliminated. I [tried this out](https://gist.github.com/baumanj/f6d587a165ed9e6f2dd3d50cb1f775c0) and it seems to work.